c10/core/DispatchKeySet.h

```diff
@@ -74,31 +252,63 @@ class DispatchKeySet final {
     return DispatchKeySet(repr_ | other.repr_);
   }
   // Perform set intersection
-  DispatchKeySet operator&(DispatchKeySet other) const {
+  constexpr DispatchKeySet operator&(DispatchKeySet other) const {
     return DispatchKeySet(repr_ & other.repr_);
   }
-  // Compute the set difference self - other
+  // Compute the set difference self - other,
+  // but ONLY for the functionality keys.
+  // Any backend bits set on self will remain unchanged.
+  // See Note [Removing keys from DispatchKeySet Only Affects Functionality Keys]
   DispatchKeySet operator-(DispatchKeySet other) const {
-    return DispatchKeySet(repr_ & ~other.repr_);
+    return DispatchKeySet(repr_ & (full_backend_mask | ~other.repr_));
```

Comment:
Hmmm I walked through this with an example that was like:

```
lets say we have 8 bits, 4 are functionality and 4 are backends.
self.repr_                   = 1110|0001
other.repr_               = 0110|0001

~other.repr_             = 1001|1110
full_backend_mask = 0000|1111
full_backend_mask | ~other.repr_ = 1001|1111
# this correctly masks out some of the functionality bits, without masking any backend bits
repr_ & (full_backend_mask | ~other.repr_) = 1000|0001
```

Do you think it's wrong / it could be written more efficiently?

Reply:
Ho ok thanks! I though the `-` operation meant something else.
Should we add to the comment that the BackendComponent from other is ignored?